BaseTools/Source/C/Makefiles/header.makefile

```diff
@@ -5,48 +5,67 @@
 # HOST_ARCH = ia32 or IA32 for IA32 build
 # HOST_ARCH = Arm or ARM for ARM build
 #
-# Copyright (c) 2007 - 2018, Intel Corporation. All rights reserved.<BR>
+# Copyright (c) 2007 - 2024, Intel Corporation. All rights reserved.<BR>
 # SPDX-License-Identifier: BSD-2-Clause-Patent
 
-EDK2_PATH ?= $(MAKEROOT)/../../..
+# Set SEP to the platform specific path seperator
+ifeq (Windows, $(findstring Windows,$(MAKE_HOST)))
+SEP:=$(shell echo \)
+else
+SEP:=/
+endif
+
+EDK2_PATH ?= $(MAKEROOT)$(SEP)..$(SEP)..$(SEP)..
 
 ifndef HOST_ARCH
-  #
-  # If HOST_ARCH is not defined, then we use 'uname -m' to attempt
-  # try to figure out the appropriate HOST_ARCH.
-  #
-  uname_m = $(shell uname -m)
-  $(info Attempting to detect HOST_ARCH from 'uname -m': $(uname_m))
-  ifneq (,$(strip $(filter $(uname_m), x86_64 amd64)))
-    HOST_ARCH=X64
-  endif
-  ifeq ($(patsubst i%86,IA32,$(uname_m)),IA32)
-    HOST_ARCH=IA32
-  endif
-  ifneq (,$(findstring aarch64,$(uname_m)))
-    HOST_ARCH=AARCH64
-  else ifneq (,$(findstring arm64,$(uname_m)))
-    HOST_ARCH=AARCH64
-  else ifneq (,$(findstring arm,$(uname_m)))
-    HOST_ARCH=ARM
-  endif
-  ifneq (,$(findstring riscv64,$(uname_m)))
-    HOST_ARCH=RISCV64
-  endif
-  ifneq (,$(findstring loongarch64,$(uname_m)))
-    HOST_ARCH=LOONGARCH64
-  endif
-  ifndef HOST_ARCH
-    $(info Could not detected HOST_ARCH from uname results)
-    $(error HOST_ARCH is not defined!)
+#
+# If HOST_ARCH is not defined, then we use 'GetGnuHostArch.py' to attempt
+# try to figure out the appropriate HOST_ARCH.
+#
+GET_GNU_HOST_ARCH_PY:=$(MAKEROOT)$(SEP)Makefiles$(SEP)GetGnuHostArch.py
+ifeq (Windows, $(findstring Windows,$(MAKE_HOST)))
+HOST_ARCH:=$(shell if defined PYTHON_COMMAND $(PYTHON_COMMAND) $(GET_GNU_HOST_ARCH_PY))
+ifeq ($(HOST_ARCH),)
+HOST_ARCH:=$(shell if not defined PYTHON_COMMAND if defined PYTHON_HOME $(PYTHON_HOME)\python.exe $(GET_GNU_HOST_ARCH_PY))
+endif
+else
+HOST_ARCH:=$(shell if command -v $(PYTHON_COMMAND) >/dev/null 1; then $(PYTHON_COMMAND) $(GET_GNU_HOST_ARCH_PY); else python $(GET_GNU_HOST_ARCH_PY); fi)
+endif
+ifeq ($(HOST_ARCH),)
+$(info HOST_ARCH detection failed.)
+undefine HOST_ARCH
+endif
+ifeq ($(HOST_ARCH),Unknown)
+$(info HOST_ARCH detection failed.)
+undefine HOST_ARCH
+endif
+endif
+ifndef HOST_ARCH
+$(error HOST_ARCH is not defined!)
+endif
+
+#Set up BaseTool binary path for Windows builds
+ifeq (Windows, $(findstring Windows,$(MAKE_HOST)))
+  ifndef BIN_PATH
+    BIN_PATH_BASE=$(MAKEROOT)$(SEP)..$(SEP)..$(SEP)Bin
+    ifeq ($(HOST_ARCH),X64)
+      BIN_PATH=$(BIN_PATH_BASE)$(SEP)Win64
+    else
+      BIN_PATH=$(BIN_PATH_BASE)$(SEP)Win32
+    endif
   endif
-  $(info Detected HOST_ARCH of $(HOST_ARCH) using uname.)
 endif
 
-CYGWIN:=$(findstring CYGWIN, $(shell uname -s))
-LINUX:=$(findstring Linux, $(shell uname -s))
-DARWIN:=$(findstring Darwin, $(shell uname -s))
-CLANG:=$(shell $(CC) --version | grep clang)
+ifeq (Windows, $(findstring Windows,$(MAKE_HOST)))

```

Comment:
Thinking about this... this probably isn't the right check to do. Really what we want to do is guard against the case that the "uname" command is not available in the current shell.

We should probably change this to:

`ifdef (,$(wildcard uname)) # Make uname exists`